mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -26,48 +38,93 @@ pub(crate) type PeerInfo = HashMap<PeerNetworkId, PeerSyncState>;
 pub(crate) struct PeerSyncState {
     pub timeline_id: u64,
     pub is_alive: bool,
-    pub broadcast_info: BroadcastInfo,
+    broadcast_info: BroadcastInfo,
 }
 
-pub(crate) struct PeerManager {
-    upstream_config: UpstreamConfig,
-    peer_info: Mutex<PeerInfo>,
-    // the upstream peer to failover to if all peers in the primary upstream network are dead
-    // the number of failover peers is limited to 1 to avoid network competition in the failover networks
-    failover_peer: Mutex<Option<PeerNetworkId>>,
+/// ACK status for a broadcasted batch of txns
+#[derive(Clone)]
+enum AckStatus {
+    /// waiting for an ACK
+    Pending,
+    /// received ACK that marks a broadcast as retriable in the future
+    Retry,
 }
 
+/// Metadata for a single broadcasted batch of txns
 #[derive(Clone)]
-pub struct BroadcastInfo {
-    // broadcasts that have not been ACK'ed for yet
-    pub sent_batches: HashMap<String, BatchInfo>,
-    // timeline IDs of all txns that need to be retried and ACKed for
-    pub total_retry_txns: BTreeSet<u64>,
-    // whether broadcasts are in backoff/backpressure mode, e.g. broadcasting at longer intervals
+struct BatchInfo {
+    /// time when this batch was last broadcast
+    pub send_time: SystemTime,
+    /// latest ACK status for this batch
+    pub ack_status: AckStatus,
+}
+
+/// Identifier for a broadcasted batch of txns
+/// For BatchId(`start_id`, `end_id`), (`start_id`, `end_id`) is the range of timeline IDs read from
+/// the core mempool timeline index that produced the txns in this batch
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd)]
+struct BatchId(u64, u64);
+
+impl BatchId {
+    /// Creates unique request id for the batch in the format "{start_id}_{end_id}"
+    fn create_request_id(&self) -> String {
+        format!("{}_{}", self.0, self.1)
+    }
+}
+
+/// Txn broadcast-related info for a given remote peer
+#[derive(Clone)]
+struct BroadcastInfo {
+    // sent broadcasts that have not yet received a success ACK
+    pub sent_batches: BTreeMap<BatchId, BatchInfo>,
```

Comment:
here, you can eliminate a BatchesInfo and have a BTreeMap<BatchId, SystemTime> for sent and BTreeSet<BatchId> for retry (or even Vecs -- I bet you the Vecs are possibly faster :P).